File: OrchidCore/src/main/java/com/eden/orchid/api/generators/GeneratorMetrics.java
package com.eden.orchid.api.generators;

import lombok.Getter;
import lombok.Setter;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.function.LongConsumer;

@Getter @Setter
public class GeneratorMetrics {

    private String key;

    private long indexingStartTime;
    private long generatingStartTime;

    private long indexingEndTime;
    private long generatingEndTime;

    private List<Long> pageGenerationTimes;

    public GeneratorMetrics(String key) {
        this.key = key;
        pageGenerationTimes = new ArrayList<>();
    }

    void startIndexing() {
        indexingStartTime = System.currentTimeMillis();
    }

    void startGenerating() {
        generatingStartTime = System.currentTimeMillis();
    }

    void stopIndexing() {
        indexingEndTime = System.currentTimeMillis();
    }

    void stopGenerating() {
        generatingEndTime = System.currentTimeMillis();
    }

    void addPageGenerationTime(long millis) {
        pageGenerationTimes.add(millis);
    }

    void addAllGenerationTimes(List<Long> millis) {
        pageGenerationTimes.addAll(millis);
    }

// Get formatted values
//----------------------------------------------------------------------------------------------------------------------

    private String makeMillisReadable(double lMillis) {
        int hours = 0;
        int minutes = 0;
        int seconds = 0;
        int millis = 0;

        String sTime;

        seconds = (int)(lMillis / 1000) % 60;
        millis = (int)(lMillis % 1000);

        if (seconds > 0) {
            minutes = (int)(lMillis / 1000 / 60) % 60;
            if (minutes > 0) {
                hours = (int)(lMillis / 1000 / 60 / 60) % 24;
                if (hours > 0) {
                    sTime = hours + "h " + minutes + "m " + seconds + "s " + millis + "ms";
                } else {
                    sTime = minutes + "m " + seconds + "s " + millis + "ms";
                }
            } else {
                sTime = seconds + "s " + millis + "ms";
            }
        } else {
            sTime = millis + "ms";
        }

        return sTime;
    }

    String getIndexingTime() {
        return makeMillisReadable(indexingEndTime - indexingStartTime);
    }

    String getGeneratingTime() {
        return makeMillisReadable(generatingEndTime - generatingStartTime);
    }

    String getMeanPageTime() {
        if(pageGenerationTimes.size() > 0) {
            return makeMillisReadable(pageGenerationTimes
                    .stream()
                    .collect(Averager::new, Averager::accept, Averager::combine)
                    .average());
        }
        else {
            return "N/A";
        }
    }

    String getMedianPageTime() {
        if(pageGenerationTimes.size() > 0) {
            pageGenerationTimes.sort(Comparator.naturalOrder());
            return makeMillisReadable(pageGenerationTimes.get((int) Math.floor(pageGenerationTimes.size()/2)));
        }
        else {
            return "N/A";
        }
    }

    static class Averager implements LongConsumer {
        private long total = 0;
        private long count = 0;

        public double average() {
            return count > 0 ? ((double) total)/count : 0;
        }
        public void accept(long i) { total += i; count++; }
        public void combine(Averager other) {
            total += other.total;
            count += other.count;
        }
    }


}

File: OrchidCore/src/main/kotlin/com/eden/orchid/api/generators/GeneratorMetrics.kt
package com.eden.orchid.api.generators

import kotlin.math.floor
import kotlin.time.Duration
import kotlin.time.DurationUnit
import kotlin.time.ExperimentalTime
import kotlin.time.milliseconds
import kotlin.time.toDuration

@OptIn(ExperimentalTime::class)
class GeneratorMetrics
@JvmOverloads
constructor(
    var key: String?,

    private var indexingStartTime: Long = Long.MAX_VALUE,
    private var generatingStartTime: Long = Long.MAX_VALUE,
    private var indexingEndTime: Long = Long.MIN_VALUE,
    private var generatingEndTime: Long = Long.MIN_VALUE
) {

    private var pageGenerationTimes = mutableListOf<Long>()

// Get formatted values
// ---------------------------------------------------------------------------------------------------------------------

    internal val indexingTime: Duration
        get() = (indexingEndTime - indexingStartTime).toDuration(DurationUnit.MILLISECONDS)

    internal val generatingTime: Duration
        get() = (generatingEndTime - generatingStartTime).toDuration(DurationUnit.MILLISECONDS)

    internal val totalTime: Duration
        get() = (generatingEndTime - indexingStartTime).toDuration(DurationUnit.MILLISECONDS)

    internal val meanPageTime: Duration
        get() = if (pageGenerationTimes.size > 0) {
            pageGenerationTimes
                .average()
                .toDuration(DurationUnit.MILLISECONDS)
        } else {
            0.milliseconds
        }

    internal val medianPageTime: Duration
        get() {
            if (pageGenerationTimes.size > 0) {
                pageGenerationTimes.sortWith(Comparator.naturalOrder())
                val medianIndex = floor((pageGenerationTimes.size / 2).toDouble()).toInt()
                return pageGenerationTimes[medianIndex].toDuration(DurationUnit.MILLISECONDS)
            } else {
                return 0.milliseconds
            }
        }

    internal val pageCount: Int
        get() = getPageGenerationTimes()!!.size

    internal fun startIndexing() {
        indexingStartTime = System.currentTimeMillis()
    }

    internal fun startGenerating() {
        generatingStartTime = System.currentTimeMillis()
    }

    internal fun stopIndexing() {
        indexingEndTime = System.currentTimeMillis()
    }

    internal fun stopGenerating() {
        generatingEndTime = System.currentTimeMillis()
    }

    internal fun addPageGenerationTime(millis: Long) {
        pageGenerationTimes.add(millis)
    }

    internal fun compose(metric: GeneratorMetrics) {
        indexingStartTime = Math.min(indexingStartTime, metric.indexingStartTime)
        indexingEndTime = Math.max(indexingEndTime, metric.indexingEndTime)
        generatingStartTime = Math.min(generatingStartTime, metric.generatingStartTime)
        generatingEndTime = Math.max(generatingEndTime, metric.generatingEndTime)
        pageGenerationTimes.addAll(metric.getPageGenerationTimes()!!)
    }

    fun getPageGenerationTimes(): List<Long>? {
        return this.pageGenerationTimes
    }

    fun setPageGenerationTimes(pageGenerationTimes: MutableList<Long>) {
        this.pageGenerationTimes = pageGenerationTimes
    }
}
